hosting/views.py

```diff
@@ -947,40 +949,46 @@ def post(self, request, *args, **kwargs):
             password=owner.password
         )
         vm_data = VirtualMachineSerializer(manager.get_vm(vm.id)).data
-        terminated = manager.delete_vm(
-            vm.id
-        )
+
+        terminated = manager.delete_vm(vm.id)
 
         if not terminated:
-            messages.error(
-                request,
-                'Error terminating VM %s' % (opennebula_vm_id)
-            )
-            return HttpResponseRedirect(self.get_success_url())
-        context = {
-            'vm': vm_data,
-            'base_url': "{0}://{1}".format(self.request.scheme,
-                                           self.request.get_host()),
-            'page_header': _('Virtual Machine Cancellation')
-        }
-        email_data = {
-            'subject': context['page_header'],
-            'to': self.request.user.email,
-            'context': context,
-            'template_name': 'vm_canceled',
-            'template_path': 'hosting/emails/',
-            'from_address': settings.DCL_SUPPORT_FROM_ADDRESS,
-        }
-        email = BaseEmail(**email_data)
-        email.send()
+            response['status'] = False
+            response['text'] = ugettext(
+                'Error terminating VM') + opennebula_vm_id
+        else:
+            for t in range(150):
+                try:
+                    manager.get_vm(self.kwargs.get('pk'))
+                except BaseException:
+                    break
+                else:
+                    sleep(2)
 
-        messages.error(
-            request,
-            _('VM %(VM_ID)s terminated successfully') % {
-                'VM_ID': opennebula_vm_id}
-        )
+            context = {
+                'vm': vm_data,
+                'base_url': "{0}://{1}".format(self.request.scheme,
+                                               self.request.get_host()),
+                'page_header': _('Virtual Machine Cancellation')
+            }
+            email_data = {
+                'subject': context['page_header'],
+                'to': self.request.user.email,
+                'context': context,
+                'template_name': 'vm_canceled',
+                'template_path': 'hosting/emails/',
+                'from_address': settings.DCL_SUPPORT_FROM_ADDRESS,
+            }
+            email = BaseEmail(**email_data)
+            email.send()
```

Comment:
Is it not possible that we send email after 150 x 2 seconds, implying that the VM is still accessible @tiwariav ?

I would rather wait for a `WrongIdError` exception which would be raised when VM is no more accessible https://github.com/ungleich/dynamicweb/blob/d30218297c8c366fc274c6a140725be52000132e/opennebula_api/models.py#L216
and send email in that specific case only.

Reply:
While testing on dev I figured that in ~60-80s the browser returns a 504 timeout response. So I had to limit the loop to 15 x 2 seconds. If the vm is not terminated in that period, the js script uses polling to check if the object exists or not. Now in this case, if the vm is terminated between two separate requests, we'll not be able to get the `vm_data` for email context.
So I think untill we start using websockets, it'll be difficult to time the emails accurately.